File: shared/src/timeline-markdown.ts
/**
 * Lightweight markdown parser for timeline detail panes.
 *
 * Output is a flat array of typed lines that each platform renders natively
 * (SwiftUI / Compose). Pure data — no rendering. Mirrors the Apple parser at
 * apple/AgentDeck/UI/Monitor/TimelineStripView.swift `TimelineMarkdownLine`
 * (which now consumes this module's output via a thin Swift port).
 *
 * Grammar (line-oriented, no inline parsing):
 *   - ``` toggles a code fence; lines inside become `code`
 *   - empty / whitespace-only line → `blank`
 *   - `# `..`### ` (1-3 hashes + space) → `heading`
 *   - `- ` or `* ` → `bullet`
 *   - `<digits>.` or `<digits>)` followed by space → `numbered`
 *   - `> ` → `quote`
 *   - anything else → `text`
 *
 * Everything inside a fence is verbatim, including any markers above.
 */

export type MarkdownLine =
  | { kind: 'blank' }
  | { kind: 'heading'; level: 1 | 2 | 3; content: string }
  | { kind: 'bullet'; content: string }
  | { kind: 'numbered'; marker: string; content: string }
  | { kind: 'quote'; content: string }
  | { kind: 'code'; content: string }
  | { kind: 'text'; content: string };

export function parseTimelineMarkdown(text: string): MarkdownLine[] {
  if (!text) return [];
  const out: MarkdownLine[] = [];
  let inCodeFence = false;

  for (const rawLine of text.split(/\r?\n/)) {
    const trimmed = rawLine.trim();

    if (trimmed.startsWith('```')) {
      inCodeFence = !inCodeFence;
      continue;
    }
    if (inCodeFence) {
      out.push({ kind: 'code', content: rawLine });
      continue;
    }
    if (trimmed.length === 0) {
      out.push({ kind: 'blank' });
      continue;
    }

    const heading = parseHeading(trimmed);
    if (heading) {
      out.push({ kind: 'heading', level: heading.level, content: heading.content });
      continue;
    }

    if (trimmed.startsWith('- ') || trimmed.startsWith('* ')) {
      out.push({ kind: 'bullet', content: trimmed.slice(2) });
      continue;
    }

    const numbered = parseNumbered(trimmed);
    if (numbered) {
      out.push({ kind: 'numbered', marker: numbered.marker, content: numbered.content });
      continue;
    }

    if (trimmed.startsWith('> ')) {
      out.push({ kind: 'quote', content: trimmed.slice(2) });
      continue;
    }

    out.push({ kind: 'text', content: rawLine });
  }

  // Match Apple's "always at least one line" contract: empty result → emit text(text)
  return out.length === 0 ? [{ kind: 'text', content: text }] : out;
}

function parseHeading(trimmed: string): { level: 1 | 2 | 3; content: string } | null {
  let level = 0;
  for (const ch of trimmed) {
    if (ch === '#') level += 1;
    else break;
  }
  if (level < 1 || level > 3) return null;
  if (trimmed.charAt(level) !== ' ') return null;
  return { level: level as 1 | 2 | 3, content: trimmed.slice(level + 1) };
}

function parseNumbered(trimmed: string): { marker: string; content: string } | null {
  const m = trimmed.match(/^(\d+)([.)])\s+(.*)$/);
  if (!m) return null;
  return { marker: `${m[1]}${m[2]}`, content: m[3] };
}

File: android/app/src/main/kotlin/dev/agentdeck/ui/timeline/TimelineMarkdown.kt
package dev.agentdeck.ui.timeline

/**
 * Parsed markdown line for the timeline detail pane.
 *
 * Mirrors the parser in `shared/src/timeline-markdown.ts` and the Apple
 * `TimelineMarkdownLine` enum in `TimelineStripView.swift`. The grammar is
 * line-oriented (no inline parsing); each parser is hand-ported and snapshot
 * tests on each platform check parity against the same fixture set.
 */
sealed class TimelineMarkdownLine {
    object Blank : TimelineMarkdownLine()
    data class Heading(val level: Int, val content: String) : TimelineMarkdownLine()
    data class Bullet(val content: String) : TimelineMarkdownLine()
    data class Numbered(val marker: String, val content: String) : TimelineMarkdownLine()
    data class Quote(val content: String) : TimelineMarkdownLine()
    data class Code(val content: String) : TimelineMarkdownLine()
    data class Plain(val content: String) : TimelineMarkdownLine()
}

/**
 * Parse `text` into a flat list of typed lines for native rendering.
 *
 * Grammar:
 *   - ``` toggles a code fence; lines inside become Code (verbatim, including markers above)
 *   - empty / whitespace-only line → Blank
 *   - 1-3 leading hashes followed by a space → Heading
 *   - "- " or "* " → Bullet
 *   - "<digits>." or "<digits>)" + space → Numbered
 *   - "> " → Quote
 *   - anything else → Plain
 */
fun parseTimelineMarkdown(text: String): List<TimelineMarkdownLine> {
    if (text.isEmpty()) return emptyList()
    val out = mutableListOf<TimelineMarkdownLine>()
    var inCodeFence = false

    for (rawLine in text.split('\n').map { it.trimEnd('\r') }) {
        val trimmed = rawLine.trim()

        if (trimmed.startsWith("```")) {
            inCodeFence = !inCodeFence
            continue
        }
        if (inCodeFence) {
            out += TimelineMarkdownLine.Code(rawLine)
            continue
        }
        if (trimmed.isEmpty()) {
            out += TimelineMarkdownLine.Blank
            continue
        }

        val heading = parseHeading(trimmed)
        if (heading != null) {
            out += TimelineMarkdownLine.Heading(heading.first, heading.second)
            continue
        }

        if (trimmed.startsWith("- ") || trimmed.startsWith("* ")) {
            out += TimelineMarkdownLine.Bullet(trimmed.substring(2))
            continue
        }

        val numbered = NUMBERED_RE.matchEntire(trimmed)
        if (numbered != null) {
            out += TimelineMarkdownLine.Numbered(
                marker = numbered.groupValues[1] + numbered.groupValues[2],
                content = numbered.groupValues[3],
            )
            continue
        }

        if (trimmed.startsWith("> ")) {
            out += TimelineMarkdownLine.Quote(trimmed.substring(2))
            continue
        }

        out += TimelineMarkdownLine.Plain(rawLine)
    }

    return if (out.isEmpty()) listOf(TimelineMarkdownLine.Plain(text)) else out
}

private val NUMBERED_RE = Regex("""^(\d+)([.)])\s+(.*)$""")

private fun parseHeading(trimmed: String): Pair<Int, String>? {
    var level = 0
    for (ch in trimmed) {
        if (ch == '#') level += 1 else break
    }
    if (level !in 1..3) return null
    if (level >= trimmed.length || trimmed[level] != ' ') return null
    return level to trimmed.substring(level + 1)
}
